dawgz/schedulers.py

```diff
@@ -54,7 +54,12 @@ async def submit(self, job: Job) -> Any:
         if job not in self.submissions:
             self.submissions[job] = asyncio.create_task(self._submit(job))
 
-        return await self.submissions[job]
+        try:
+            result = await self.submissions[job]
+        except Exception as error:
+            result = error
+
+        return result
```

Comment:
This is problematic as it would catch literally all exceptions. For instance, if the `run('sbatch')` call fails, it should break immediately the scheduling.

Reply:
The `DependencyNeverSatisfiedException` is propagated. For instance
```python
from dawgz import job, after, waitfor, require, ensure, schedule, leafs

@job(array=2)
def a(i: int):
    if i == 1:
        raise Exception()
    print('a')

@after(a, status='failure')
@job
def a_fail():
    print('a fail')

@after(a, status='success')
@job
def c():
    print('c')

schedule(*leafs(a), backend='local')
```
outputs
```a
a
a fail
Traceback (most recent call last):
  File "/home/joeri/Workspace/dawgz/dawgz/schedulers.py", line 119, in _submit
    return await asyncio.gather(*(
  File "/home/joeri/Workspace/dawgz/dawgz/utils.py", line 29, in to_thread
    return await loop.run_in_executor(None, func_call)
  File "/home/joeri/Anaconda/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/joeri/Workspace/dawgz/dawgz/workflow.py", line 104, in call
    result = f(*args)
  File "/home/joeri/Workspace/dawgz/test.py", line 8, in a
    raise Exception()
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joeri/Workspace/dawgz/test.py", line 32, in <module>
    results = schedule(*leafs(a), backend='local')
  File "/home/joeri/Workspace/dawgz/dawgz/schedulers.py", line 35, in schedule
    return asyncio.run(scheduler.gather(*jobs))
  File "/home/joeri/Anaconda/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/joeri/Anaconda/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/home/joeri/Workspace/dawgz/dawgz/schedulers.py", line 49, in gather
    raise result
  File "/home/joeri/Workspace/dawgz/dawgz/schedulers.py", line 58, in submit
    result = await self.submissions[job]
  File "/home/joeri/Workspace/dawgz/dawgz/schedulers.py", line 101, in _submit
    raise DependencyNeverSatisfiedException(f'aborting job {job}') from result
dawgz.schedulers.DependencyNeverSatisfiedException: aborting job c
```
Which is the expected behaviour. The same holds for `run('sbatch')`.